src/libraries/System.Net.Http/tests/StressTests/HttpStress/StressClient.cs

```diff
@@ -450,7 +450,7 @@ public void PrintFailureTypes()
         }
 
 
-        private class StructuralEqualityComparer<T> : IEqualityComparer<T> where T : IStructuralEquatable
+        private class StructuralEqualityComparer<T> : IEqualityComparer<T> where T : class, IStructuralEquatable
```

Comment:
I assume you were getting a warning about GetHashCode? Instead of adding a class constraint, it would have [Disallow null] on the T argument to GetHashCode.

Reply:
No, I was getting errors in `Equals(T? left, T? right)`.

Reply:
That's then because you're not using a new enough compiler.

Reply:
Actually, no, this should have been working fine. Something else is wrong. There shouldn't be any need for the class constraint here.

What is the exact error? What SDK are you using?

Reply:
[This is the error on CI builds](https://dev.azure.com/dnceng/public/_build/results?buildId=774062&view=logs&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=ac0b0b0f-051f-52f7-8fb3-a7e384b0dde9&l=72).

Reply:
On CI, test are running in Docker, so the SDK is provided by the respective Docker image
- On Windows - `mcr.microsoft.com/dotnet/nightly/sdk:5.0-nanoserver-1809`
- On Linux - `mcr.microsoft.com/dotnet/nightly/sdk:5.0-buster-slim`

Reply:
Those errors are about the security tests not http tests, no?

Reply:
There are also [errors on HTTP tests](https://dev.azure.com/dnceng/public/_build/results?buildId=774060&view=logs&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=ac0b0b0f-051f-52f7-8fb3-a7e384b0dde9&l=68). The message is different, but seems related.

Reply:
That's different. Try it with a nightly to get a newer compiler.

Reply:
But, it already uses base image with nightly SDK. Or, is there some other way to get the right Docker image with nightly SDK?

Reply:
I just tried reverting this change and building inside a container based on `mcr.microsoft.com/dotnet/nightly/sdk:5.0-nanoserver-1809`, but it again failed with the same error.
```
StressClient.cs(455,32): error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint. [C:\app\HttpStress.csproj]

StressClient.cs(455,41): error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint. [C:\app\HttpStress.csproj]
```

Reply:
I have SDK Version: 5.0.100-rc.1.20418.3 installed from https://github.com/dotnet/installer/blob/master/README.md#installers-and-binaries.

Building master, HttpStress builds successfully with no changes:
```
D:\repos\runtime\src\libraries\System.Net.Http\tests\StressTests\HttpStress [master]> dotnet build /t:rebuild
Microsoft (R) Build Engine version 16.8.0-preview-20406-01+e8c17c128 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  HttpStress -> D:\repos\runtime\src\libraries\System.Net.Http\tests\StressTests\HttpStress\bin\Debug\net5.0\HttpStress.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.08
D:\repos\runtime\src\libraries\System.Net.Http\tests\StressTests\HttpStress [master]>
```

SslStress does correctly fail to build with the warning:
```
D:\repos\runtime\src\libraries\System.Net.Security\tests\StressTests\SslStress\Utils\ErrorAggregator.cs(115,25): warning CS8767: Nullability of reference types in type of parameter 'left' of 'bool StructuralEqualityComparer<T>.Equals(T left, T right)' doesn't match implicitly implemented member 'bool IEqualityComparer<T>.Equals(T x, T y)' (possibly because of nullability attributes). [D:\repos\runtime\src\libraries\System.Net.Security\tests\StressTests\SslStress\SslStress.csproj]
```
The fix for that is changing L115 from:
```
public bool Equals(T left, T right) => left.Equals(right, StructuralComparisons.StructuralEqualityComparer);
```
to something like:
```C#
public bool Equals(T? left, T? right) => left is null ? right is null : left.Equals(right, StructuralComparisons.StructuralEqualityComparer);
```

As for CI, I don't actually know where that docker image pulls bits from.  @MichaelSimons, do you know?

Reply:
Looking at the [CI failure](https://dev.azure.com/dnceng/public/_build/results?buildId=774062&view=logs&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=ac0b0b0f-051f-52f7-8fb3-a7e384b0dde9&l=72) @alnikola referenced earlier, it appears CI is using the official Docker nightly images (the latest 5.0 image).  These are produced from the [nightly branch of the dotnet-docker repo](https://github.com/dotnet/dotnet-docker/tree/nightly).  

My initial reaction is that it seems broken that the product build/test is using these images.  I naively think you would want to use the same toolset the build is using.  They way things are currently defined (e.g referencing the latest nightly 5.0 tag), there is no control as to which version of 5.0 is being used.  I expect the product builds should be using a deterministic version of the product.

Reply:
@MichaelSimons  Stress pipelines are normally being run nightly, so referencing the official Docker nightly images is justified from cost/benefit perspective.
Could you please tell me when those images were updated last time? When could we expect them to be updated onto SDK 5.0.100-rc.1.20418.3 or newer?

>They way things are currently defined (e.g referencing the latest nightly 5.0 tag), there is no control as to which version of 5.0 is being used.

It's intentional to run stress tests on the latest 5.0 version. We want to see how tests behavior changes due to changes in production code.

Reply:
Because 5.0 Preview 8 contains security fixes and therefore there are no public builds available, the nightly images haven't been updated since P7.  They won't get updated to the RC channel until P8 is released.  You can see this yourself by looking at the checked in Dockerfiles [here](https://github.com/dotnet/dotnet-docker/tree/nightly/src).  You can also see the image publish dates and find the links to the Dockerfile the images were built with by looking at Tag Listing section of the [Docker Hub repos](https://hub.docker.com/_/microsoft-dotnet-nightly-sdk).

If I may offer a suggestion, consider modify the tests/Dockerfile to echo out which SDK/Runtime version it is testing with given the use of the floating Docker tag.  This seems like beneficial info to have when investigating failures.

Reply:
Thanks for the explanation. I will add echoing out SDK version to Docker files (Runtime version is already logged by the test apps).

Reply:
Checked with latest SDK locally. Nullability failure are gone now, but **Ssl**Stress build still fails on `OperatingSystem.IsWindows`, so I'll revert generic type constraint changes, but keep the OS check's one.